lib/snippets/kubernetes-deploy

```diff
@@ -75,90 +203,375 @@ class KubernetesDeploy
+    if namespaced
+      raise FatalDeploymentError, "Namespace missing for namespaced command" unless @namespace
+      args.push("--namespace=#{@namespace}")
+    end
+    KubernetesDeploy.logger.debug Shellwords.join(args)
     out, err, st = Open3.capture3(*args)
-    logger.debug(out.shellescape)
-    logger.warn(err) unless st.success?
+    KubernetesDeploy.logger.debug(out.shellescape)
+    KubernetesDeploy.logger.warn(err) unless st.success?
     [out.chomp, err.chomp, st]
   end
 
-  def logger
+  def phase_heading(phase_name)
+    @current_phase += 1
+    heading = "Phase #{@current_phase}: #{phase_name}"
+    padding = (100.0 - heading.length)/2
+    KubernetesDeploy.logger.info("")
+    KubernetesDeploy.logger.info("#{'-' * padding.floor}#{heading}#{'-' * padding.ceil}")
+  end
+
+  def log_green(msg)
+    STDOUT.puts "\033[0;32m#{msg}\x1b[0m\n" # green
+  end
+
+  def self.logger
     @logger ||= begin
       l = Logger.new(STDOUT)
       l.level = ENV["DEBUG"] ? Logger::DEBUG : Logger::INFO
       l.formatter = proc do |severity, _datetime, _progname, msg|
         case severity
-        when "FATAL" then "\033[0;31m[#{severity}]\t#{msg}\x1b[0m\n"
-        when "ERROR", "WARN" then "\033[0;31m[#{severity}]\t#{msg}\x1b[0m\n"
-        when "INFO" then "\033[0;36m#{msg}\x1b[0m\n"
+        when "FATAL" then "\033[0;31m[#{severity}]\t#{msg}\x1b[0m\n" # red
+        when "ERROR", "WARN" then "\033[0;33m[#{severity}]\t#{msg}\x1b[0m\n" # yellow
+        when "INFO" then "\033[0;36m#{msg}\x1b[0m\n" # blue
         else "[#{severity}]\t#{msg}\n"
         end
       end
       l
     end
   end
+
+  class KubernetesResource
+    attr_reader :name, :namespace, :file
+    attr_writer :type
+
+    # Source: Rails
+    # https://github.com/rails/rails/blob/b292b76c2dd0f04fb090d49b90716a0e6037b41a/activesupport/lib/active_support/core_ext/class/subclasses.rb#L22-L29
+    def self.descendants
+      descendants = []
+      ObjectSpace.each_object(singleton_class) do |k|
+        next if k.singleton_class?
+        descendants.unshift k unless k == self
+      end
+      descendants
+    end
+
+    def self.handled_type
+      name.split('::').last.downcase
+    end
+
+    def self.for_type(type, name, namespace, file)
+      if subclass = descendants.find { |subclass| subclass.handled_type == type }
+        subclass.new(name, namespace, file)
+      else
+        self.new(name, namespace, file).tap { |r| r.type = type }
+      end
+    end
+
+    def initialize(name, namespace, file)
+      # subclasses must also set these
+      @name, @namespace, @file = name, namespace, file
+    end
+
+    def id
+      "#{type}/#{name}"
+    end
+
+    def sync
+      log_status(status_data)
+    end
+
+    def failed?
+      false
+    end
+
+    def succeeded?
+      KubernetesDeploy.logger.warn("Don't know how to monitor resources of type #{type}. Assuming #{id} deployed successfully.")
+      true
+    end
+
+    def exists?
```

Comment:
It's the latter (the base class will be used by any resource of a type there's no specific handler for). Will do.